src/components/ListView.js

```diff
@@ -99,25 +105,90 @@ export const ListView = props => h(Element, Object.assign({
 
 export const listView = ({
   component: (state, actions) => {
+    // TODO: Ability to deselect with control and shift key
+    const createSelection = (index) => state.selectedIndex.indexOf(index) === -1
+      ? [...state.selectedIndex, index]
+      : state.selectedIndex;
+
+    /**
+     * Creates a range of indexes from start to end
+     * @param {Number} start
+     * @param {Number} end
+     * @return {Array}
+     */
+    const createSelectionRange = (start, end) => {
+      // Swaps start and end if start is greater than end
+      if (start > end) {
+        [start, end] = [end, start];
+      }
+
+      const indices = [
+        ...state.selectedIndex,
+        // Generates a range of indexes from start to end
+        ...Array.from({length: end - start + 1}, (_, i) => i + start)
+      ];
+
+      // Remove duplicates from the array
+      return [...new Set(indices)];
+    };
+
+    const getSelection = (index, ev) => {
+      const selected = state.multiselect
+        ? (ev.shiftKey
+          ? createSelectionRange(state.previousSelectedIndex, index)
+          : ev.ctrlKey
+            ? createSelection(index)
+            : [index])
+        : index;
+
+      const data = state.multiselect
+        ? selected.map((item) => state.rows[item].data)
+        : state.rows[selected].data;
+
+      // Store the previous index in the state to use for calculating the
+      // range if the shift key is pressed
+      if (state.multiselect) {
+        state.previousSelectedIndex = index;
```

Comment:
Mutating the state object is not technically allowed.

Ideally you wanna return a new state from this function, then you can call an action to update the state.

```js
const {selected, data, previousSelectedIndex} = getSelection(index, ev);
// ...
actions.setPreviousSelectedIndex(previousSelectedIndex);

// ...
setPreviousSelectedIndex: previousSelectedIndex => ({previousSelectedIndex})
```

Reply:
I agree with the premise and I'll add this in either tonight or tomorrow, but can you elaborate on "not allowed"? Is it just like discouraged because it's bad practice or what?

Reply:
Most UI frameworks and libraries provides state in some way or another, and in order to make the app "reactive", you have to use some kind of state interface for this to happen.

If not, state changes aren't picked up correctly and the app will misbehave.

The updating of state cause a re-render, and if the state has updated without the UI catching up, you're in for a bad time!

---

The equivlient to this in React using hooks would be:

```js
const [previousSelectedIndex, setPreviousSelectedIndex] = useState():

// ...
function someKindOfEvent() {
  const selection = getSelection(/* ... */);
  setPreviousSelectedIndex(selection.previousSelectedIndex);
}
```

Or some other familiar pattern is (like in Redux):

```js
const initialState = {
  previousSelectedIndex: undefined,
}

//

function setPreviousSelectedIndex(currentState, value) {
 return {
    ...currentState,
    previousSelectedIndex: value
 };
}
```

Reply:
Gotcha, thanks! Will be sure to fix a push tonight/tomorrow :)

Reply:
So, basically. Use some kind of "storage" to update the state. Don't mutate the state object directly :)